File: Lecture 2/FinalCode - Java/app/src/main/java/com/inovaeg/androidtrainingpharos2020/MainActivity.java
package com.inovaeg.androidtrainingpharos2020;

import android.Manifest;
import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.net.Uri;
import android.os.Bundle;
import android.provider.MediaStore;
import android.util.Log;
import android.widget.ImageView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;

public class MainActivity extends AppCompatActivity {
    private String TAG = this.getClass().getSimpleName();
    private int locationPermissionCode = 1;
    private int cameraPermissionCode = 2;
    private int capturePhotoCode = 2;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Log.d(TAG, "onCreate: ");

        findViewById(R.id.bt_take_permission).setOnClickListener(v -> {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA)
                    != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.CAMERA},
                        cameraPermissionCode);
            } else {
                showToast("you have permission to access the camera");
            }

        });

        findViewById(R.id.bt_start_camera).setOnClickListener(v -> {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA)
                    != PackageManager.PERMISSION_GRANTED) {
                showToast("please request permission first");
            } else {
                Intent takePictureIntent = new Intent(MediaStore.ACTION_IMAGE_CAPTURE);
                if (takePictureIntent.resolveActivity(getPackageManager()) != null) {
                    startActivityForResult(takePictureIntent, capturePhotoCode);
                }

            }

        });
        findViewById(R.id.bt_check_location).setOnClickListener(v -> {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
                    != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.ACCESS_FINE_LOCATION},
                        locationPermissionCode);
            } else {
                addLocationListener();
            }

        });

    }

    @SuppressLint("MissingPermission")
    private void addLocationListener() {
        showToast("adding location Listener");
        LocationManager locationManager = (LocationManager) getSystemService(Context.LOCATION_SERVICE);
        locationManager
                .requestLocationUpdates(LocationManager.GPS_PROVIDER, 10000, 10, new LocationListener() {
                    @Override
                    public void onLocationChanged(@NonNull Location location) {
                        Uri gmmIntentUri = Uri.parse("geo:" + location.getLatitude() + "," + location.getLongitude());
                        Intent mapIntent = new Intent(Intent.ACTION_VIEW, gmmIntentUri);
                        mapIntent.setPackage("com.google.android.apps.maps");
                        if (mapIntent.resolveActivity(getPackageManager()) != null) {
                            startActivity(mapIntent);
                        }
                        locationManager.removeUpdates(this);

                    }
                });
    }

    private void showToast(@NonNull String message) {
        Toast.makeText(this, message, Toast.LENGTH_LONG).show();
    }

    @Override
    protected void onStart() {
        super.onStart();
        Log.d(TAG, "onStart: ");
    }

    @Override
    protected void onResume() {
        super.onResume();
        Log.d(TAG, "onResume: ");
    }

    @Override
    protected void onPause() {
        super.onPause();
        Log.d(TAG, "onPause: ");
    }

    @Override
    protected void onStop() {
        super.onStop();
        Log.d(TAG, "onStop: ");
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Log.d(TAG, "onDestroy: ");
    }

    @Override
    protected void onRestart() {
        super.onRestart();
        Log.d(TAG, "onRestart: ");
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == cameraPermissionCode) {
            if (checkGrantedPermission(grantResults)) {
                showToast("you have taken Camera permission");
            } else showToast("camera permission was denied");
        } else if (requestCode == locationPermissionCode) {
            if (checkGrantedPermission(grantResults)) {
                showToast("you have taken location permission");
                addLocationListener();
            } else showToast("location permission was denied");
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == capturePhotoCode && resultCode == RESULT_OK && data != null && data.getExtras() != null) {
            Bundle extras = data.getExtras();
            Bitmap imageBitmap = (Bitmap) extras.get("data");
            ((ImageView) findViewById(R.id.imageView)).setImageBitmap(imageBitmap);
        }

    }

    private boolean checkGrantedPermission(@NonNull int[] grantedPermission) {
        return grantedPermission.length > 0 &&
                grantedPermission[0] == PackageManager.PERMISSION_GRANTED;
    }
}

File: Lecture 2/FinalCode - Kotlin/app/src/main/java/com/inovaeg/androidtrainingpharos2020/MainActivity.kt
package com.inovaeg.androidtrainingpharos2020

import android.Manifest
import android.annotation.SuppressLint
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.util.Log
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {
    private val TAG = this.javaClass.simpleName
    private val locationPermissionCode = 1
    private val cameraPermissionCode = 2
    private val capturePhotoCode = 2
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        Log.d(TAG, "onCreate: ")
        bt_take_permission.setOnClickListener {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA)
                    != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.CAMERA),
                        cameraPermissionCode)
            } else {
                showToast("you have permission to access the camera")
            }
        }
        bt_start_camera.setOnClickListener {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.CAMERA)
                    != PackageManager.PERMISSION_GRANTED) {
                showToast("please request permission first")
            } else {
                val takePictureIntent = Intent(MediaStore.ACTION_IMAGE_CAPTURE)
                if (takePictureIntent.resolveActivity(packageManager) != null) {
                    startActivityForResult(takePictureIntent, capturePhotoCode)
                }
            }
        }
        bt_check_location.setOnClickListener {
            if (ContextCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION)
                    != PackageManager.PERMISSION_GRANTED) {
                ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.ACCESS_FINE_LOCATION),
                        locationPermissionCode)
            } else {
                addLocationListener()
            }
        }
    }

    @SuppressLint("MissingPermission")
    private fun addLocationListener() {
        showToast("adding location Listener")
        (getSystemService(LOCATION_SERVICE) as LocationManager)
                .apply {
                    requestLocationUpdates(LocationManager.GPS_PROVIDER, 10000, 10f, object : LocationListener {
                        override fun onLocationChanged(location: Location) {
                            val gmmIntentUri = Uri.parse("geo:" + location.latitude + "," + location.longitude)
                            val mapIntent = Intent(Intent.ACTION_VIEW, gmmIntentUri)
                            mapIntent.setPackage("com.google.android.apps.maps")
                            if (mapIntent.resolveActivity(packageManager) != null) {
                                startActivity(mapIntent)
                            }
                            this@apply.removeUpdates(this)
                        }
                    })
                }

    }

    private fun showToast(message: String) {
        Toast.makeText(this, message, Toast.LENGTH_LONG).show()
    }

    override fun onStart() {
        super.onStart()
        Log.d(TAG, "onStart: ")
    }

    override fun onResume() {
        super.onResume()
        Log.d(TAG, "onResume: ")
    }

    override fun onPause() {
        super.onPause()
        Log.d(TAG, "onPause: ")
    }

    override fun onStop() {
        super.onStop()
        Log.d(TAG, "onStop: ")
    }

    override fun onDestroy() {
        super.onDestroy()
        Log.d(TAG, "onDestroy: ")
    }

    override fun onRestart() {
        super.onRestart()
        Log.d(TAG, "onRestart: ")
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode == cameraPermissionCode) {
            if (checkGrantedPermission(grantResults)) {
                showToast("you have taken Camera permission")
            } else showToast("camera permission was denied")
        } else if (requestCode == locationPermissionCode) {
            if (checkGrantedPermission(grantResults)) {
                showToast("you have taken location permission")
                addLocationListener()
            } else showToast("location permission was denied")
        }
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == capturePhotoCode && resultCode == RESULT_OK && data != null && data.extras != null) {
            val extras = data.extras
            val imageBitmap = extras!!["data"] as Bitmap?
            imageView.setImageBitmap(imageBitmap)
        }
    }

    private fun checkGrantedPermission(grantedPermission: IntArray): Boolean {
        return grantedPermission.isNotEmpty() &&
                grantedPermission[0] == PackageManager.PERMISSION_GRANTED
    }
}